packages/editor/src/components/rich-text/format-toolbar/index.native.js

```diff
@@ -0,0 +1,205 @@
+		this.onKeyDown = this.onKeyDown.bind( this );
+		this.onChangeLinkValue = this.onChangeLinkValue.bind( this );
+		this.toggleLinkSettingsVisibility = this.toggleLinkSettingsVisibility.bind( this );
+		this.setLinkTarget = this.setLinkTarget.bind( this );
+	}
+
+	onKeyDown( event ) {
+		if ( event.keyCode === ESCAPE ) {
+			const link = this.props.formats.link;
+			const isAddingLink = link && link.isAdding;
+			if ( isAddingLink ) {
+				event.stopPropagation();
+				if ( ! link.value ) {
+					this.dropLink();
+				} else {
+					this.props.onChange( { link: { ...link, isAdding: false } } );
+				}
+			}
+		}
+		if ( [ LEFT, DOWN, RIGHT, UP, BACKSPACE, ENTER ].indexOf( event.keyCode ) > -1 ) {
+			stopKeyPropagation( event );
+		}
+	}
+
+	static getDerivedStateFromProps( props, state ) {
+		if ( state.selectedNodeId !== props.selectedNodeId ) {
+			return computeDerivedState( props );
+		}
+
+		return null;
+	}
+
+	onChangeLinkValue( value ) {
+		this.setState( { linkValue: value } );
+	}
+
+	toggleFormat( format ) {
+		return () => {
+			this.props.onChange( {
+				[ format ]: ! this.props.formats[ format ],
+			} );
+		};
+	}
+
+	toggleLinkSettingsVisibility() {
+		this.setState( ( state ) => ( { settingsVisible: ! state.settingsVisible } ) );
+	}
+
+	setLinkTarget( opensInNewWindow ) {
+		this.setState( { opensInNewWindow } );
+		if ( this.props.formats.link && ! this.props.formats.link.isAdding ) {
+			this.props.onChange( { link: {
+				value: this.props.formats.link.value,
+				target: opensInNewWindow ? '_blank' : null,
+				rel: opensInNewWindow ? 'noreferrer noopener' : null,
+			} } );
+		}
+	}
+
+	addLink() {
+		this.setState( { linkValue: '' } );
+		this.props.onChange( { link: { isAdding: true } } );
+	}
+
+	dropLink() {
+		this.props.onChange( { link: null } );
+		this.setState( { linkValue: '' } );
+	}
+
+	editLink( event ) {
+		event.preventDefault();
+		this.props.onChange( { link: { ...this.props.formats.link, isAdding: true } } );
+		this.setState( { linkValue: this.props.formats.link.value } );
+	}
+
+	submitLink( event ) {
+		event.preventDefault();
+		const value = prependHTTP( this.state.linkValue );
+		this.props.onChange( { link: {
+			isAdding: false,
+			target: this.state.opensInNewWindow ? '_blank' : null,
+			rel: this.state.opensInNewWindow ? 'noreferrer noopener' : null,
+			value,
+		} } );
+
+		this.setState( { linkValue: value } );
+		if ( ! this.props.formats.link.value ) {
+			this.props.speak( __( 'Link added.' ), 'assertive' );
+		}
+	}
+
+	isFormatActive( format ) {
+		return this.props.formats[ format ] && this.props.formats[ format ].isActive;
+	}
+
+	render() {
+		const { enabledControls = DEFAULT_CONTROLS, customControls = [] } = this.props;
+		// const { linkValue, settingsVisible, opensInNewWindow } = this.state;
+		// const isAddingLink = formats.link && formats.link.isAdding;
+
+		const toolbarControls = FORMATTING_CONTROLS.concat( customControls )
```

Comment:
It looks like all the difference between native and web in this file boils down to the `render` implementation. Can we extract another component instead and get rid of all this code duplication which quickly will become hard to maintain?

Reply:
There is some import that are commented out as you can see on the top of the file. Should we go ahead and and create empty native version of those?

Reply:
I would extract `render` method to its own component first and then see if there is anything else to override in the original file. My assumption is that all overrides happen inside render.